0018-ircp.md

```diff
@@ -0,0 +1,234 @@
+# ADR 0018: Inter-Runtime Communication Protocol (IRCP)
+
+## Component
+
+Oasis Core, Oasis SDK
+
+## Changelog
+
+- 2022-10-13: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Currently each runtime is its own relatively isolated environment. It can
+interact with the consensus layer via [incoming] and [outgoing] messages, but
+a standardized way to communicate with other parallel runtimes sharing the same
+consensus layer is missing.
+
+This ADR proposes the Inter-Runtime Communication Protocol (IRCP) which enables
+any two runtimes sharing the same consensus layer to securely communicate by
+exchanging standardized packets. It also proposes modifications to the runtime
+P2P network to support seamless relaying of messages without the need for any
+external processes.
+
+[incoming]: 0011-incoming-runtime-messages.md
+[outgoing]: 0003-consensus-runtime-token-transfer.md
+
+## Decision
+
+The following new concepts are introduced by this proposal:
+
+* Each runtime can signal that a **connection** should be established _to_
+  another runtime sharing the same consensus layer. For each unique destination
+  runtime only one connection may be established.
+
+* In the context of an established connection the runtime may establish one or
+  more **channels**. Each channel can be used to send and receive ordered
+  packets.
+
+The transport of packets does not require any new interactions with the
```

Comment:
It is a bit unclear how this would work on other nodes that do not run both runtimes (where do they get the packets from?). If packets are not written to the consensus layer or elsewhere, how do new nodes verify which packets were received in the past and in what way they affected the state?

Reply:
This will be more clear once the protocol flow sections are written. But to answer briefly:
* Packets are transported via P2P, similar to transactions (see the _Transport_ section), and are verified using Merkle proofs and light clients (see the _Authentication_ section).
* All nodes sharing the consensus layer have a trusted view of consensus light headers. Since the consensus layer state stores all state roots of all runtimes, any node can obtain the state root at given height of any runtime.
* Verifiable packet batches (generated by the source runtime which does have access to the entire runtime state) contain Merkle proofs, rooted at the runtime's state root, proving a certain leaf is stored in source runtime's storage.
* Usually the leaf to be proven is the outgoing packet being stored in the source runtime's send queue.
* This proves (to the destination runtime) that the packet was indeed sent by the given runtime.
* Ordering is verified by tracking sequence numbers.